### DISCUSSION TITLE
Show whole trade session for 1Day chart

### DISCUSSION THREAD

**Discussion by HarshaKosala** [tradingview/lightweight-charts#1150] [Q&A]
Hi,

Is there a way to show the whole trade session in the chart area and fill the candles from the left to right when trade happens. Currently new candles are drawing from the right side. Instead is there a way to let grow the chart from left to right by showing the whole trade session to the user in the chart viewport.  

Thanks

**Comment by SlicedSilver** [comment]
Sounds like you want to control the scroll position and zoom for the time scale.

For an introduction to the time scale we have this documentation page: https://tradingview.github.io/lightweight-charts/docs/time-scale 

and here is the API for the time scale: https://tradingview.github.io/lightweight-charts/docs/api/interfaces/ITimeScaleApi

You would most likely need to use the `setVisibleLogicalRange` method or the `setVisibleRange ` method.

Hope that this helps.

**Reply by HarshaKosala** [reply]
Hi @SlicedSilver,
setVisibleRange() method not set the range if candles are not available until 'to' time. in my case there is no chart data until that time for now but when time goes chart is drawing. 
setVisibleLogicalRange() also not resulting the intended behavior.  
Is there any other way to do this ?

Thanks

**Reply by SlicedSilver** [reply]
I threw together a quick example of what I think you are trying to achieve: https://jsfiddle.net/msilverwood/qu3n4cgs/3/

**Note:** It's also important to set the `shiftVisibleRangeOnNewBar` ([docs](https://tradingview.github.io/lightweight-charts/docs/api/interfaces/TimeScaleOptions#shiftvisiblerangeonnewbar)) property to `false` for the time scale.

How the example works:
* Created some sample data which I split into two parts: what to show immediately, and what to 'reveal' at a later stage.
* Set the data for the initial data to the candlestick series.
* Get the current logical visible range.
* set the `to` value for the logical range to the current value plus the length of what will be revealed.
* Animate the reveal of the new data, where each data point is added with the `update` method on the series.

**Reply by HarshaKosala** [reply]
Hi @SlicedSilver 

shiftVisibleRangeOnNewBar () method seems not working. 

![image](https://user-images.githubusercontent.com/19972882/185175573-beb3f672-9ec0-4919-abd8-92d508f8cef8.png)
this.chartObj.timeScale().fitContent();

![image](https://user-images.githubusercontent.com/19972882/185175767-42339d0a-98fa-4788-94fb-ece0090ef7ae.png)

It's not adding any empty space to the right side.
Chart is updating using a real time data feed

**Reply by SlicedSilver** [reply]
Perhaps if you could put together a small JSFiddle code sample illustrating the problem then I could determine if there is a bug or which part of your code to change.

I wouldn't adjust the visibleLogicalRange within a `subscribeVisibleLogicalRangeChange`. I would recommend using `setVisibleLogicalRange` once when the chart is first populated with data, and then not adjust it anymore (so that the user can scroll around).

Also if you are using `fitContent()` then that will adjust the time scale range.

**Reply by HarshaKosala** [reply]
Hi @SlicedSilver 
Initially i didn't used subscribeVisibleLogicalRangeChange() method. But setVisibleLogicalRange () didn't work initially. That's why i moved to subscribe method.

**Reply by SlicedSilver** [reply]
Have a look at the JSFiddle example I posted earlier. The order of the method calls is important. Perhaps you were evoking `getVisibleLogicalRange ` or `setVisibleLogicalRange ` before using `setData`.

**Reply by HarshaKosala** [reply]
Hi @SlicedSilver 
Below are the code lines 

```
this.series.setData(data);
const timeScale = this.chartObj.timeScale();
timeScale.applyOptions({
	shiftVisibleRangeOnNewBar: false,
});
const visibleLogicalRange = timeScale.getVisibleLogicalRange();
if (visibleLogicalRange) {
	visibleLogicalRange.to += 50;
	timeScale.setVisibleLogicalRange(visibleLogicalRange);
}
this.chartObj.timeScale().fitContent();
this.subscribeForRealTimeChartData(exgSym, symbolKey); // update the chart real time using a datafeed. 
       this.series.update(realTimeDataObj.point);
```

**Reply by SlicedSilver** [reply]
I would remove the `this.chartObj.timeScale().fitContent();` line. 

`fitContent` will adjust the visible range to fit all the data currently in the series. It doesn't make sense to call this after you've manually edited the range by using `setVisibleLogicalRange` because those changes will be lost.

You probably want to have actually have the `fitContent` call to happen earlier (before `timeScale.getVisibleLogicalRange `).

Changing your code (not tested):
```js
this.series.setData(data);
const timeScale = this.chartObj.timeScale();
timeScale.applyOptions({
	shiftVisibleRangeOnNewBar: false,
});

this.chartObj.timeScale().fitContent(); // Move here

const visibleLogicalRange = timeScale.getVisibleLogicalRange();
if (visibleLogicalRange) {
	visibleLogicalRange.to += 50;
	timeScale.setVisibleLogicalRange(visibleLogicalRange);
}

this.subscribeForRealTimeChartData(exgSym, symbolKey); // update the chart real time using a datafeed. 
       this.series.update(realTimeDataObj.point);
```

**Reply by HarshaKosala** [reply]
Hi @SlicedSilver 

Followed exact same code but still no result

**Reply by SlicedSilver** [reply]
If you could create a JSFiddle (or equivalent) of your issue then I could take a look and try assist you further.

Otherwise, I would again suggest looking at the JSFiddle example linked above (https://jsfiddle.net/msilverwood/qu3n4cgs/3/), and make sure that you are on the latest release of the library (3.8.0)

**Comment by Wash01** [comment]
thanks for asking 

> Hi,
> 
> Is there a way to show the whole trade session in the chart area and fill the candles from the left to right when trade happens. Currently new candles are drawing from the right side. Instead is there a way to let grow the chart from left to right by showing the whole trade session to the user in the chart viewport.
> 
> Thanks
